### ISSUE TITLE
Display the source file name

### ISSUE THREAD

**Issue by nedbat** [inducer/pudb#170]
When I don't know which file I am looking at, I use "L" to see the file name.  Pudb could display the name in the title of the source pane.

**Comment by inducer** [comment]
That's a whole row of characters down the tube though. I appreciate every row not occupied by UI, especially on size-constrained displays. Plus I'm not that often disoriented about where I am, but then I'm mostly debugging my own code. I could see it as an option, maybe.

**Comment by nedbat** [comment]
I wasn't going to use a new row, I would add it to the very top row, which has a lot of space left.

**Comment by asmeurer** [comment]
The top row seems pretty full if you have a smaller terminal (isn't the default on most systems 80 characters wide?). The row the says "command line" is pretty empty, though.

**Comment by inducer** [comment]
Agree that the top row is pretty full. That "command line" line should probably be removed (rather than cluttered up more).

**Comment by asmeurer** [comment]
Then again, if you run PuDB fullscreen (like I do and I'm guessing @nedbat does) there is plenty of room at the top. So it could be added there, but only visible for people willing to make their terminals wide enough.

**Comment by inducer** [comment]
That'd be good, but it's a little nontrivial to achieve in Urwid. (i.e. you're writing a custom `Widget` at that point)

**Comment by nedbat** [comment]
This might be controversial, but I don't think the top line needs as much help as it offers.  "PuDB - ?:help" would be enough.

**Comment by inducer** [comment]
I could go along with that.

**Comment by asmeurer** [comment]
I mean only visible in the sense of what it already does now (truncate the top line).

**Comment by asmeurer** [comment]
Oh nevermind, it doesn't truncate. I misremembered.

**Comment by timblaktu** [comment]
I agree with @nedbat about preferring the top line to only have the "? - help" tip, and to include the current file name/path. I would suggest by default it display N chars from the end of the abspath of the current filename, where N is determined by an algorithm fed by a $COLUMNS from the shell, plus sidebar_width from the current pudb.conf settings. I prefer this solution to peer into the stack window for current filename.

**Comment by ndevenish** [comment]
> That's a whole row of characters down the tube though
Even as an option would be great - jumping around in large codebases I frequently have to open the L dialog just to see where I am, would be great if that could be seen at a glance

**Comment by dkao1978** [comment]
Having an option for this would be amazing. Anyone who's navigating up and down the stack will appreciate this. This is especially useful if you're trying to debug inheritance problems in a large code base.

"L" does not work in the Stack so you have to go back and forth between the stack and the main window to see which file you're looking at.

**Comment by TimothyBramlett** [comment]
This is fantastic idea and love the filename being displayed!

**Comment by qhuy4119** [comment]
I am experimenting with this and at a glance it looks quite easy. 
The content of the top row is in this [caption](https://github.com/inducer/pudb/blob/b087d646cb532af1bc3396e4d133967eb49fee2a/pudb/debugger.py#L2641) variable. We can get the current filename from `DebuggerUI.source_code_provider.get_source_identifier()` if I'm not mistaken.

From there other things like dynamic caption length computed from $COLUMNS, this feature being optional,... can be done quickly.

**Comment by inducer** [comment]
If it doesn't cost vertical real estate (which can be precious in row-constrained terminals), I'm OK with having this added.

**Comment by qhuy4119** [comment]
I got it to display the full path to current source file. However, if the path is too long to be displayed in 1 row, it will wrap automatically, creating more rows. If we somehow have enough more space (by resizing the window, for example), a Ctrl-L will redraw it into 1 row.
![Screenshot_20210713_195215](https://user-images.githubusercontent.com/30318745/125455908-04ff9b86-9756-4197-abb9-fe77282166de.png)

What do we want to do?

**Comment by inducer** [comment]
The wrapping is kind of what I wanted to avoid. `urwid.Text` has a `wrap="ellipsis"` mode, which may help enforce the single-line requirement. It'll probably clip the most interesting part though (the actual filename). Maybe we can start by *only* showing the filename? AFAICT, a better solution would require us to write a custom widget.

**Comment by qhuy4119** [comment]
I just tried all the wrap options of `urwid.Text` and none looks promising. 
I think we can:
- show only the filename
- show as much of the full path as possible on 1 line("/.../dirA/subdir/foo.py"). This will help providing more context. We can start from root or from an arbitrary dir ("dirA/subdir/foo.py" instead)

**Comment by inducer** [comment]
> none looks promising.

Why? Can you explain?

> show only the filename

I think this with `ellipsis` clipping would be my preferred "cheap" solution.

> show as much of the full path as possible on 1 line

How would we know?

> have a config in the config file for this

Nah. This is a solvable problem. We don't need to involve the user.

**Comment by qhuy4119** [comment]
> > none looks promising.
> 
> Why? Can you explain?
> 
> > show only the filename
> 
> I think this with `ellipsis` clipping would be my preferred "cheap" solution.
> 
> > show as much of the full path as possible on 1 line
> 
> How would we know?
> 
> > have a config in the config file for this
> 
> Nah. This is a solvable problem. We don't need to involve the user.

1. In the wrap options: `space` and `any` will wrap. `clip` and `ellipsis` will cut out the ending, which is what we actually want.
2. Like above
3. I think we can calculate this based on current available horizontal space
4. My reason is that in most(or many) cases, people are working with enough screen space to display full path. So we can have a default and still let users decide based on their use cases.

Point 3 and 4 may be because of my overthinking though. Just let me know what we decide.

**Comment by qhuy4119** [comment]
I updated the PR so that:
if there is enough width: display the full path, 
else display the longest possible path. 

If the terminal size changes, PuDB will update the path automatically.

**Comment by BIAOXYZ** [comment]
I agree this. Without this functionality, I have to use `ctrl + x` then manually print `__file__` to see the full path of current file- -

**Comment by inducer** [comment]
You can just use `Shift + L`, too.

**Comment by guoyejun** [comment]
> * show as much of the full path as possible on 1 line("/.../dirA/subdir/foo.py"). This will help providing more context. We can start from root or from an arbitrary dir ("dirA/subdir/foo.py" instead)

good idea imo.

and I think we'd not display it in the title of the source pane, because it will overwrite the help information especially ?:help, I used to learn pudb by pressing ?

another option is to display the filename (with as much of the full path as possible) at the line between the source pane and command line pane by replacing "Command line: [Ctrl-X]", just like the behavior of cgdb.

**Comment by kwmiebach** [comment]
It is currently not shown in the main window to save lines. Type [shift-L] to see the complete path, then [right arrow][enter] to go back.
